File: go-sdk/worker/call.go
// Package worker provides functionality for calling methods on the scheduler.
package worker

import (
	"crypto/rand"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"reflect"

	"github.com/go-enols/go-server/go-sdk/scheduler"
)

const (
	TaskStatusError = "error"
)

// Call 执行远程方法调用，支持可选的加密参数
// 如果提供了 encryptionKey，则使用加密方式调用
func Call(host, method string, params, out interface{}, encryptionKey ...string) error {
	// 检查out参数是否为指针类型
	if out != nil {
		outType := reflect.TypeOf(out)
		if outType.Kind() != reflect.Ptr {
			return fmt.Errorf("参数out必须是指针类型或nil")
		}
	}
	client := scheduler.NewSchedulerClient(host)

	// 如果提供了加密密钥，使用加密方式调用
	if len(encryptionKey) > 0 && encryptionKey[0] != "" {
		// 生成随机盐值
		saltBig, err := rand.Int(rand.Reader, big.NewInt(900000))
		if err != nil {
			return fmt.Errorf("生成随机盐值失败: %v", err)
		}
		salt := int(saltBig.Int64()) + 100000 // 生成6位随机数
		res, err := client.ExecuteEncrypted(method, encryptionKey[0], salt, params)
		if err != nil {
			return err
		}
		if res.Status == TaskStatusError {
			return errors.New(string(res.Result))
		}
		// 轮询结果
		result, err := client.GetResultEncrypted(res.TaskID, encryptionKey[0], salt)
		if err != nil {
			return err
		}
		if result.Status == TaskStatusError {
			return errors.New(string(result.Result))
		}
		if out != nil {
			return json.Unmarshal(result.Result, out)
		}
		return nil
	}

	// 普通调用
	res, err := client.Execute(method, params)
	if err != nil {
		return err
	}
	if res.Status == TaskStatusError {
		return errors.New(string(res.Result))
	}
	// 轮询结果
	result, err := client.GetResult(res.TaskID)
	if err != nil {
		return err
	}

	if result.Status == TaskStatusError {
		return errors.New(string(result.Result))
	}
	if out != nil {
		return json.Unmarshal(result.Result, out)
	}
	return nil
}

File: node-sdk/worker/call.js
const SchedulerClient = require('../scheduler/client');

/**
 * 简单的RPC调用函数
 * @param {string} host - 调度器地址
 * @param {string} method - 方法名
 * @param {any} params - 参数
 * @returns {Promise<any>} 返回结果
 */
async function call(host, method, params) {
  const client = new SchedulerClient(host);

  const res = await client.execute(method, params);

  // 直接获取结果，execute 只返回 taskId 和 status
  const result = await client.getResult(res.taskId);

  if (result.status === 'error') {
    throw new Error(result.result);
  }

  return result.result;
}

/**
 * 加密的RPC调用函数
 * @param {string} host - 调度器地址
 * @param {string} method - 方法名
 * @param {string} key - 加密密钥
 * @param {number} salt - 加密盐值
 * @param {any} params - 参数
 * @returns {Promise<any>} 返回结果
 */
async function callEncrypted(host, method, key, salt, params) {
  const client = new SchedulerClient(host);
  
  try {
    // 执行加密任务
    const response = await client.executeEncrypted(method, key, salt, params);
    const taskId = response.taskId;
    
    // 轮询结果，设置超时
    const timeout = 30000; // 30秒
    const startTime = Date.now();
    
    while (Date.now() - startTime < timeout) {
      const result = await client.getResult(taskId);
      
      if (result.status === 'completed') {
        return result.result;
      } else if (result.status === 'error') {
        throw new Error(result.result);
      }
      
      // 等待后再次轮询
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    throw new Error(`加密任务 ${taskId} 在 ${timeout}ms 内未完成`);
  } finally {
    // 客户端清理自动处理
  }
}

module.exports = { call, callEncrypted };
// 向后兼容
module.exports.default = call;
